KEP-3178-related iptables rule fixups

danwinship opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Noticed while rereading KEP-3178 in preparation for pushing it to beta.

1. I had claimed we were going to make kube-proxy create its own copy of the "anti-martian-packet" rule that kubelet creates, but we never did this.
2. I had noticed a while back that the removal of the `-j KUBE-MARK-DROP` rules made the load balancer firewall chains look sort of ambiguous; when a packet falls off the end of the chain, it's not clear to someone reading the rules what is going to happen to it. So now we just append a gratuitous extra comment rule explaining.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3178
```

/sig network

danwinship committed:
Add a comment-only rule to the end of KUBE-FW-* chains

With the removal of the "-j KUBE-MARK-DROP" rules, the firewall chains
end rather ambiguously. Add a comment-only rule explaining what will
happen.

danwinship committed:
fix indentation of iptables dumps in some test cases

(Especially, use tabs rather than spaces.)

danwinship committed:
Duplicate the "anti-martian-packet" rule in kube-proxy

This rule was mistakenly added to kubelet even though it only applies
to kube-proxy's traffic. We do not want to remove it from kubelet yet
because other components may be depending on it for security, but we
should make kube-proxy output its own rule rather than depending on
kubelet.

aojea commented:
interesting, this breaks kind jobs https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/114470/pull-kubernetes-e2e-kind/1602775061524123648/artifacts/kind-worker/pods/kube-system_kube-proxy-65qhg_9f56e4ad-3922-4d64-87d7-c200d9a75f72/kube-proxy/0.log

kind has to use the route_localhost for DNS IIRC
/cc @bentheelder

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -942,6 +945,22 @@ func (proxier *Proxier) syncProxyRules() {
+		// iptables rule mitigates. Note that for historical reasons, kubelet
+		// creates a similar rule (in KUBE-FIREWALL), but that should eventually
+		// go away.
+		proxier.filterRules.Write(
```
why the kubelet rule works and this breaks kind?

danwinship replied:
I doubt this is the part of the PR that's breaking kind; it's mostly identical to the kubelet rule except that the kubelet rule does `! --ctstate RELATED,ESTABLISHED,DNAT`, while this rule does `--ctstate NEW`, so it's actually dropping _less_.

aojea replied:
that is the only new rule, no? I didn't check carefully and most sure I'm wrong

danwinship replied:
ok, it _is_ that rule that's breaking it...

aojea replied:
is it the order or what? this in example of kind rules

```
docker exec -it kind-control-plane iptables-save
# Generated by iptables-save v1.8.7 on Wed Dec 14 17:09:15 2022
*mangle
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:KUBE-IPTABLES-HINT - [0:0]
:KUBE-KUBELET-CANARY - [0:0]
:KUBE-PROXY-CANARY - [0:0]
COMMIT
# Completed on Wed Dec 14 17:09:15 2022
# Generated by iptables-save v1.8.7 on Wed Dec 14 17:09:15 2022
*filter
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:KUBE-EXTERNAL-SERVICES - [0:0]
:KUBE-FIREWALL - [0:0]
:KUBE-FORWARD - [0:0]
:KUBE-KUBELET-CANARY - [0:0]
:KUBE-NODEPORTS - [0:0]
:KUBE-PROXY-CANARY - [0:0]
:KUBE-PROXY-FIREWALL - [0:0]
:KUBE-SERVICES - [0:0]
-A INPUT -m conntrack --ctstate NEW -m comment --comment "kubernetes load balancer firewall" -j KUBE-PROXY-FIREWALL
-A INPUT -m comment --comment "kubernetes health check service ports" -j KUBE-NODEPORTS
-A INPUT -m conntrack --ctstate NEW -m comment --comment "kubernetes externally-visible service portals" -j KUBE-EXTERNAL-SERVICES
-A INPUT -j KUBE-FIREWALL
-A FORWARD -m conntrack --ctstate NEW -m comment --comment "kubernetes load balancer firewall" -j KUBE-PROXY-FIREWALL
-A FORWARD -m comment --comment "kubernetes forwarding rules" -j KUBE-FORWARD
-A FORWARD -m conntrack --ctstate NEW -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A FORWARD -m conntrack --ctstate NEW -m comment --comment "kubernetes externally-visible service portals" -j KUBE-EXTERNAL-SERVICES
-A OUTPUT -m conntrack --ctstate NEW -m comment --comment "kubernetes load balancer firewall" -j KUBE-PROXY-FIREWALL
-A OUTPUT -m conntrack --ctstate NEW -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A OUTPUT -j KUBE-FIREWALL
-A KUBE-EXTERNAL-SERVICES -p tcp -m comment --comment "default/nodeport-service has no local endpoints" -m addrtype --dst-type LOCAL -m tcp --dport 31964 -j DROP
-A KUBE-FIREWALL ! -s 127.0.0.0/8 -d 127.0.0.0/8 -m comment --comment "block incoming localnet connections" -m conntrack ! --ctstate RELATED,ESTABLISHED,DNAT -j DROP
-A KUBE-FIREWALL -m comment --comment "kubernetes firewall for dropping marked packets" -m mark --mark 0x8000/0x8000 -j DROP
-A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP
-A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x4000/0x4000 -j ACCEPT
-A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
COMMIT
# Completed on Wed Dec 14 17:09:15 2022
# Generated by iptables-save v1.8.7 on Wed Dec 14 17:09:15 2022
*nat
:PREROUTING ACCEPT [124:8370]
:INPUT ACCEPT [88:5957]
:OUTPUT ACCEPT [3366568:201994080]
:POSTROUTING ACCEPT [3366634:201999241]
:DOCKER_OUTPUT - [0:0]
:DOCKER_POSTROUTING - [0:0]
:KIND-MASQ-AGENT - [0:0]
:KUBE-EXT-QATUYZLPU4I4HE4S - [0:0]
:KUBE-KUBELET-CANARY - [0:0]
:KUBE-MARK-DROP - [0:0]
:KUBE-MARK-MASQ - [0:0]
:KUBE-NODEPORTS - [0:0]
:KUBE-POSTROUTING - [0:0]
:KUBE-PROXY-CANARY - [0:0]
:KUBE-SEP-6E7XQMQ4RAYOWTTM - [0:0]
:KUBE-SEP-7VBYVD4YJPPXAPRS - [0:0]
:KUBE-SEP-CZUVPU4NP442FRNG - [0:0]
:KUBE-SEP-JPCU5XURIL5XWZWM - [0:0]
:KUBE-SEP-PUHFDAMRBZWCPADU - [0:0]
:KUBE-SEP-SF3LG62VAE5ALYDV - [0:0]
:KUBE-SEP-WXWGHGKZOCNYRYI7 - [0:0]
:KUBE-SEP-ZP3FB6NMPNCO4VBJ - [0:0]
:KUBE-SEP-ZXMNUKOKXUTL2MK2 - [0:0]
:KUBE-SERVICES - [0:0]
:KUBE-SVC-ERIFXISQEP7F7OF4 - [0:0]
:KUBE-SVC-JD5MR3NA4I4DYORP - [0:0]
:KUBE-SVC-NPX46M4PTMTKRN6Y - [0:0]
:KUBE-SVC-QATUYZLPU4I4HE4S - [0:0]
:KUBE-SVC-TCOU7JCQXEZGVUNU - [0:0]
-A PREROUTING -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A PREROUTING -d 192.168.8.1/32 -j DOCKER_OUTPUT
-A OUTPUT -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A OUTPUT -d 192.168.8.1/32 -j DOCKER_OUTPUT
-A POSTROUTING -m comment --comment "kubernetes postrouting rules" -j KUBE-POSTROUTING
-A POSTROUTING -d 192.168.8.1/32 -j DOCKER_POSTROUTING
-A POSTROUTING -m addrtype ! --dst-type LOCAL -m comment --comment "kind-masq-agent: ensure nat POSTROUTING directs all non-LOCAL destination traffic to our custom KIND-MASQ-AGENT chain" -j KIND-MASQ-AGENT
-A DOCKER_OUTPUT -d 192.168.8.1/32 -p tcp -j DNAT --to-destination 127.0.0.11:36147
-A DOCKER_OUTPUT -d 192.168.8.1/32 -p udp -j DNAT --to-destination 127.0.0.11:48082
-A DOCKER_POSTROUTING -s 127.0.0.11/32 -p tcp -j SNAT --to-source 192.168.8.1:53
-A DOCKER_POSTROUTING -s 127.0.0.11/32 -p udp -j SNAT --to-source 192.168.8.1:53
-A KIND-MASQ-AGENT -d 10.244.0.0/16 -m comment --comment "kind-masq-agent: local traffic is not subject to MASQUERADE" -j RETURN
-A KIND-MASQ-AGENT -m comment --comment "kind-masq-agent: outbound traffic is subject to MASQUERADE (must be last in chain)" -j MASQUERADE
-A KUBE-EXT-QATUYZLPU4I4HE4S -s 10.244.0.0/16 -m comment --comment "pod traffic for default/nodeport-service external destinations" -j KUBE-SVC-QATUYZLPU4I4HE4S
-A KUBE-EXT-QATUYZLPU4I4HE4S -m comment --comment "masquerade LOCAL traffic for default/nodeport-service external destinations" -m addrtype --src-type LOCAL -j KUBE-MARK-MASQ
-A KUBE-EXT-QATUYZLPU4I4HE4S -m comment --comment "route LOCAL traffic for default/nodeport-service external destinations" -m addrtype --src-type LOCAL -j KUBE-SVC-QATUYZLPU4I4HE4S
-A KUBE-MARK-DROP -j MARK --set-xmark 0x8000/0x8000
-A KUBE-MARK-MASQ -j MARK --set-xmark 0x4000/0x4000
-A KUBE-NODEPORTS -p tcp -m comment --comment "default/nodeport-service" -m tcp --dport 31964 -j KUBE-EXT-QATUYZLPU4I4HE4S
-A KUBE-POSTROUTING -m mark ! --mark 0x4000/0x4000 -j RETURN
-A KUBE-POSTROUTING -j MARK --set-xmark 0x4000/0x0
-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE --random-fully
-A KUBE-SEP-6E7XQMQ4RAYOWTTM -s 10.244.0.3/32 -m comment --comment "kube-system/kube-dns:dns" -j KUBE-MARK-MASQ
-A KUBE-SEP-6E7XQMQ4RAYOWTTM -p udp -m comment --comment "kube-system/kube-dns:dns" -m udp -j DNAT --to-destination 10.244.0.3:53
-A KUBE-SEP-7VBYVD4YJPPXAPRS -s 10.244.1.3/32 -m comment --comment "default/nodeport-service" -j KUBE-MARK-MASQ
-A KUBE-SEP-7VBYVD4YJPPXAPRS -p tcp -m comment --comment "default/nodeport-service" -m tcp -j DNAT --to-destination 10.244.1.3:8080
-A KUBE-SEP-CZUVPU4NP442FRNG -s 10.244.2.2/32 -m comment --comment "default/nodeport-service" -j KUBE-MARK-MASQ
-A KUBE-SEP-CZUVPU4NP442FRNG -p tcp -m comment --comment "default/nodeport-service" -m tcp -j DNAT --to-destination 10.244.2.2:8080
-A KUBE-SEP-JPCU5XURIL5XWZWM -s 192.168.8.2/32 -m comment --comment "default/kubernetes:https" -j KUBE-MARK-MASQ
-A KUBE-SEP-JPCU5XURIL5XWZWM -p tcp -m comment --comment "default/kubernetes:https" -m tcp -j DNAT --to-destination 192.168.8.2:6443
-A KUBE-SEP-PUHFDAMRBZWCPADU -s 10.244.0.4/32 -m comment --comment "kube-system/kube-dns:metrics" -j KUBE-MARK-MASQ
-A KUBE-SEP-PUHFDAMRBZWCPADU -p tcp -m comment --comment "kube-system/kube-dns:metrics" -m tcp -j DNAT --to-destination 10.244.0.4:9153
-A KUBE-SEP-SF3LG62VAE5ALYDV -s 10.244.0.4/32 -m comment --comment "kube-system/kube-dns:dns-tcp" -j KUBE-MARK-MASQ
-A KUBE-SEP-SF3LG62VAE5ALYDV -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp" -m tcp -j DNAT --to-destination 10.244.0.4:53
-A KUBE-SEP-WXWGHGKZOCNYRYI7 -s 10.244.0.4/32 -m comment --comment "kube-system/kube-dns:dns" -j KUBE-MARK-MASQ
-A KUBE-SEP-WXWGHGKZOCNYRYI7 -p udp -m comment --comment "kube-system/kube-dns:dns" -m udp -j DNAT --to-destination 10.244.0.4:53
-A KUBE-SEP-ZP3FB6NMPNCO4VBJ -s 10.244.0.3/32 -m comment --comment "kube-system/kube-dns:metrics" -j KUBE-MARK-MASQ
-A KUBE-SEP-ZP3FB6NMPNCO4VBJ -p tcp -m comment --comment "kube-system/kube-dns:metrics" -m tcp -j DNAT --to-destination 10.244.0.3:9153
-A KUBE-SEP-ZXMNUKOKXUTL2MK2 -s 10.244.0.3/32 -m comment --comment "kube-system/kube-dns:dns-tcp" -j KUBE-MARK-MASQ
-A KUBE-SEP-ZXMNUKOKXUTL2MK2 -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp" -m tcp -j DNAT --to-destination 10.244.0.3:53
-A KUBE-SERVICES -d 10.96.50.77/32 -p tcp -m comment --comment "default/nodeport-service cluster IP" -m tcp --dport 80 -j KUBE-SVC-QATUYZLPU4I4HE4S
-A KUBE-SERVICES -d 10.96.0.10/32 -p udp -m comment --comment "kube-system/kube-dns:dns cluster IP" -m udp --dport 53 -j KUBE-SVC-TCOU7JCQXEZGVUNU
-A KUBE-SERVICES -d 10.96.0.10/32 -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-SVC-ERIFXISQEP7F7OF4
-A KUBE-SERVICES -d 10.96.0.10/32 -p tcp -m comment --comment "kube-system/kube-dns:metrics cluster IP" -m tcp --dport 9153 -j KUBE-SVC-JD5MR3NA4I4DYORP
-A KUBE-SERVICES -d 10.96.0.1/32 -p tcp -m comment --comment "default/kubernetes:https cluster IP" -m tcp --dport 443 -j KUBE-SVC-NPX46M4PTMTKRN6Y
-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
-A KUBE-SVC-ERIFXISQEP7F7OF4 ! -s 10.244.0.0/16 -d 10.96.0.10/32 -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp -> 10.244.0.3:53" -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-ZXMNUKOKXUTL2MK2
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp -> 10.244.0.4:53" -j KUBE-SEP-SF3LG62VAE5ALYDV
-A KUBE-SVC-JD5MR3NA4I4DYORP ! -s 10.244.0.0/16 -d 10.96.0.10/32 -p tcp -m comment --comment "kube-system/kube-dns:metrics cluster IP" -m tcp --dport 9153 -j KUBE-MARK-MASQ
-A KUBE-SVC-JD5MR3NA4I4DYORP -m comment --comment "kube-system/kube-dns:metrics -> 10.244.0.3:9153" -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-ZP3FB6NMPNCO4VBJ
-A KUBE-SVC-JD5MR3NA4I4DYORP -m comment --comment "kube-system/kube-dns:metrics -> 10.244.0.4:9153" -j KUBE-SEP-PUHFDAMRBZWCPADU
-A KUBE-SVC-NPX46M4PTMTKRN6Y ! -s 10.244.0.0/16 -d 10.96.0.1/32 -p tcp -m comment --comment "default/kubernetes:https cluster IP" -m tcp --dport 443 -j KUBE-MARK-MASQ
-A KUBE-SVC-NPX46M4PTMTKRN6Y -m comment --comment "default/kubernetes:https -> 192.168.8.2:6443" -j KUBE-SEP-JPCU5XURIL5XWZWM
-A KUBE-SVC-QATUYZLPU4I4HE4S ! -s 10.244.0.0/16 -d 10.96.50.77/32 -p tcp -m comment --comment "default/nodeport-service cluster IP" -m tcp --dport 80 -j KUBE-MARK-MASQ
-A KUBE-SVC-QATUYZLPU4I4HE4S -m comment --comment "default/nodeport-service -> 10.244.1.3:8080" -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-7VBYVD4YJPPXAPRS
-A KUBE-SVC-QATUYZLPU4I4HE4S -m comment --comment "default/nodeport-service -> 10.244.2.2:8080" -j KUBE-SEP-CZUVPU4NP442FRNG
-A KUBE-SVC-TCOU7JCQXEZGVUNU ! -s 10.244.0.0/16 -d 10.96.0.10/32 -p udp -m comment --comment "kube-system/kube-dns:dns cluster IP" -m udp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SVC-TCOU7JCQXEZGVUNU -m comment --comment "kube-system/kube-dns:dns -> 10.244.0.3:53" -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-6E7XQMQ4RAYOWTTM
-A KUBE-SVC-TCOU7JCQXEZGVUNU -m comment --comment "kube-system/kube-dns:dns -> 10.244.0.4:53" -j KUBE-SEP-WXWGHGKZOCNYRYI7
COMMIT
```

aojea replied:
is it solved now, what was the problem?

aojea replied:
see the comment now https://github.com/kubernetes/kubernetes/pull/114470#issuecomment-1367593437

danwinship committed:
Add iptablesKubeletJumpChains to iptables proxier

Some of the chains kube-proxy creates are also created by kubelet; we
need to ensure that those chains exist but we should not delete them
in CleanupLeftovers().

danwinship commented:
@aojea fixed this a while back but just repushed again; I don't know exactly what was wrong with the rule before, but given that it took us a few tries to get the existing rule exactly right, it's probably best to just duplicate it _exactly_, so now we just create exactly the same rule as kubelet does, just like with do with the masquerade rules.

danwinship commented:
/retest-required

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1360,8 +1360,12 @@ func (proxier *Proxier) syncProxyRules() {
 			// corresponding KUBE-PROXY-FIREWALL rule.
+			proxier.natRules.Write(
+				"-A", string(fwChain),
+				"-m", "comment", "--comment", fmt.Sprintf(`"other traffic to %s will be dropped by KUBE-PROXY-FIREWALL"`, svcPortNameString),
```
this is mostly curiosity, does this rules got evaluated at all? is there any. performance implications on having them?

danwinship replied:
I don't know if it actually gets evaluated in data plane processing at all... but it should not have any interesting performance impact.

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -373,6 +373,13 @@ var iptablesJumpChains = []iptablesJumpChain{
+
+// Duplicates of chains created in pkg/kubelet/kubelet_network_linux.go; we create these
+// on startup but do not delete them in CleanupLeftovers.
+var iptablesKubeletJumpChains = []iptablesJumpChain{
```
so this is a bugfix, isn't it?

danwinship replied:
Technically, yes; right now, if you run `kube-proxy --proxy-mode=iptables --cleanup`, then it will delete a chain that kubelet "cares about", and kubelet will eventually recreate it. But the chain in question is one that kubelet doesn't actually _use_, so nothing will break.

(I wouldn't have bothered to fix this except that it seemed bad to have kube-proxy delete `KUBE-FIREWALL` on `--cleanup`, but then if I was fixing that, it seemed wrong to not fix `KUBE-POSTROUTING` too...)

aojea commented:
/lgtm

/hold

@thockin may want to take a look

aojea commented:
/assign @thockin
